qutip/solver/result.py

```diff
@@ -970,3 +980,110 @@ def runs_photocurrent(self):
                 for i in range(self.num_c_ops)
             ])
         return measurements
+
+
+class NmmcTrajectoryResult(McTrajectoryResult):
+    """
+    Result class used by the :class:`qutip.NonMarkovianMCSolver` for single
+    trajectories. Additionally stores the trace of the state along the
+    trajectory.
+    """
+
+    def __init__(self, e_ops, options, solver, *args, **kwargs):
+        super().__init__(e_ops, options, *args, **kwargs)
+
+        self._solver = solver
+        self.trace = []
+
+    # This gets called during the Monte-Carlo simulation of the associated
+    # completely positive master equation. To obtain the state of the actual
+    # system, we simply multiply the provided state with the current martingale
+    # before storing it / computing expectation values.
+    def add(self, t, state):
+        if isket(state):
+            state = ket2dm(state)
```

Comment:
Is converting a ket state to a dm the right thing to do here? If MCSolve, wouldn't it be better to rather do something like:
```
if isket(state):
    state = state * np.sqrt(mu)
else:
   state = state * mu
```
so that the kind of state is preserved in the results?

Reply:
Unfortunately, that would not work. The method only works with density matrices, not with ket states. The reason is that mu may be negative and in that case, `ket2dm(ket * sqrt(mu))` is not the same as `mu * ket2dm(ket)`.

I will add a note in the docs to document this behavior.